// @typegoose/typegoose/lib/esm/hooks.js
import { DecoratorKeys } from './internal/constants';
import { assertion, getName } from './internal/utils';
import { logger } from './logSettings';
// Note: TSDoc for the hooks can't be added without adding it to *every* overload
const hooks = {
    pre(...args) {
        return (target) => addToHooks(target, 'pre', args);
    },
    post(...args) {
        return (target) => addToHooks(target, 'post', args);
    }
};
/**
 * Add a hook to the hooks Array
 * @param target Target Class
 * @param hookType What type is it
 * @param args All Arguments, that should be passed-throught
 */
function addToHooks(target, hookType, args) {
    var _a, _b;
    // Convert Method to array if only a string is provided
    const methods = Array.isArray(args[0]) ? args[0] : [args[0]];
    assertion(typeof args[1] === 'function', new TypeError(`"${getName(target)}.${hookType}.${methods.join(' ')}"'s function is not a function!`));
    const func = args[1];
    logger.info('Adding hooks for "[%s]" to "%s" as type "%s"', methods.join(','), getName(target), hookType);
    for (const method of methods) {
        switch (hookType) {
            case 'post':
                const postHooks = Array.from((_a = Reflect.getMetadata(DecoratorKeys.HooksPost, target)) !== null && _a !== void 0 ? _a : []);
                postHooks.push({ func, method });
                Reflect.defineMetadata(DecoratorKeys.HooksPost, postHooks, target);
                break;
            case 'pre':
                const preHooks = Array.from((_b = Reflect.getMetadata(DecoratorKeys.HooksPre, target)) !== null && _b !== void 0 ? _b : []);
                preHooks.push({ func, method });
                Reflect.defineMetadata(DecoratorKeys.HooksPre, preHooks, target);
                break;
        }
    }
}
export const pre = hooks.pre;
export const post = hooks.post;
// Export it PascalCased
export const Pre = hooks.pre;
export const Post = hooks.post;
